addons/sale_project/models/project.py

```diff
@@ -370,20 +371,49 @@ def _get_profitability_sale_order_items_domain(self, domain=None):
     def _get_revenues_items_from_sol(self, domain=None, with_action=True):
         sale_line_read_group = self.env['sale.order.line'].sudo()._read_group(
             self._get_profitability_sale_order_items_domain(domain),
-            ['product_id'],
+            ['product_id', 'is_downpayment'],
             ['id:array_agg', 'untaxed_amount_to_invoice:sum', 'untaxed_amount_invoiced:sum'],
         )
         display_sol_action = with_action and len(self) == 1 and self.user_has_groups('sales_team.group_sale_salesman')
         revenues_dict = {}
         total_to_invoice = total_invoiced = 0.0
+        data = []
+        sequence_per_invoice_type = self._get_profitability_sequence_per_invoice_type()
         if sale_line_read_group:
-            sols_per_product = {
-                product.id: (
-                    untaxed_amount_to_invoice,
-                    untaxed_amount_invoiced,
-                    ids,
-                ) for product, ids, untaxed_amount_to_invoice, untaxed_amount_invoiced in sale_line_read_group
-            }
+            sols_per_product = {}
+            downpayment_amount_invoiced = 0
+            downpayment_sol_ids = []
+            for res in sale_line_read_group:
+                if res[1]:
+                    downpayment_amount_invoiced += res[4]
+                    downpayment_sol_ids += res[2]
+                else:
+                    sols_per_product[res[0].id] = (
+                        res[3],
+                        res[4],
+                        res[2],
+                    )
```

Comment:
May I ask you to "destructure" the tuple to name each element in that tuple. 😇 

```suggestion
            for product_id, is_downpayment, sol_ids, total_amount_to_invoice, total_amount_invoiced in sale_line_read_group:
                if is_downpayment:
                    downpayment_amount_invoiced += res[4]
                    downpayment_sol_ids += res[2]
                else:
                    sols_per_product[product_id.id] = (
                        total_amount_to_invoice,
                        total_amount_invoiced,
                        sol_ids,
                    )
```

It will inscrease the readability of your code. 🙂